### PR TITLE
allow React.ReactElement in register option's message

### LINKED ISSUE #1266: Allow React components to be returned as `validate` rule error message
Hello,

**Is your feature request related to a problem? Please describe.**
My issue is that the `validate` method only allows to return a `bool` or a `string`, and returning a React component will not validate as an error. Being able to set a React component as an error message would allow for more complex and useful (UI-wise) error handling.
As an example, consider a password input with a dynamic rule checking for its constraints (required characters, length, etc.). The `validate` method works very well with this kind of complex rule checks, but I would also like to be able to give instant feedback as the user types a password, via a react component instead of a string.

I would also like to point out that setting a component as an error message in the other validation methods works (I tried with `pattern`) (
```js
register({
    pattern({
        value: /regex/,
        message: component
    }),
})
```
).

**Describe the solution you'd like**
I'd like the `validate` method to accept a React component as a possible return type, to be handled as a message error. 

**Describe alternatives you've considered**
None.

**Additional context**
To give you a more concrete example, here is a condensed snippet of what I am trying to do:

```js
register({
    validate: customValidate, 
})
```

```js
customValidate: value => {
    const constraints = {
        hasMinLen: value.length >= 8,
        hasUppercase: /([A-Z]+)/.test(value),
        hasLowercase: /([a-z]+)/.test(value),
        hasNumber: /([0-9]+)/.test(value),
    };

     return (
         Object.values(constraints).every(constraint => constraint) || (
            <span>
                Le mot de passe doit contenir au moins:
                <div className="tags">
                    {!constraints.hasMinLen && (
                        <span className="tag is-danger">8 caractères</span>
                    )}
                    {!constraints.hasLowercase && (
                        <span className="tag is-danger">1 minuscule</span>
                    )}
                    {!constraints.hasUppercase && (
                        <span className="tag is-danger">1 majuscule</span>
                    )}
                    {!constraints.hasNumber && (
                        <span className="tag is-danger">1 chiffre</span>
                    )}
                </div>
            </span>
         )
    );
}
```

Thank you very much for your attention!

### PR DESCRIPTION
**Pull request by kotarella1110**
Todo:

- [x] add unit test
- [x] improve type

Usage:

```tsx
register({
  required: <CustomErrorMessage message="required" />,
  max: {
    value: 1000,
    message: <CustomErrorMessage message="max" />,
  },
  min: {
    value: 10,
    message: <CustomErrorMessage message="min" />,
  },
  validate: {
    test: value => value === '100' || <CustomErrorMessage  message="validate" />,
  },
})
```

```tsx
<ErrorMessage
  name="test"
  errors={errors}
  message={<CustomErrorMessage message="ErrorMessage component" />}
/>
```

```tsx
// Single Error
setError(
  "test",
  "singleError",
  <CustomErrorMessage message="single error" />
)
// Single Errors
setError("test", {
  single1: <CustomErrorMessage message="single error1" />,
  single2: <CustomErrorMessage message="single error2" />
})
// Multiple Errors
setError([
  {
    type: "multiple",
    name: "test1",
    message: <CustomErrorMessage message="multiple error" />
  },
  {
    type: "multiple",
    name: "test2",
    message: <CustomErrorMessage message="multiple error" />
  }
])
```

Test it out: https://codesandbox.io/s/react-hook-form-issue-1266-khtse

close #1266

### CODE REVIEW

File: src/contextTypes.ts [hunk 1]
```diff
@@ -66,11 +67,11 @@ export type FormContextValues<FormValues extends FieldValues = FieldValues> = {
   ): FieldValue<FormValues> | DeepPartial<FormValues> | string | undefined;
   setError(name: ManualFieldError<FormValues>[]): void;
   setError(name: FieldName<FormValues>, type: MultipleFieldErrors): void;
-  setError(name: FieldName<FormValues>, type: string, message?: string): void;
+  setError(name: FieldName<FormValues>, type: string, message?: Message): void;
```
Comment (bluebill1049) [inline src/contextTypes.ts]: oh nice! you fixed `setError` as well!

File: src/logic/getValidateError.ts [hunk 2]
```diff
@@ -1,16 +1,18 @@
...
-import { FieldError, ValidateResult, Ref } from '../types';
+import { FieldError, ValidateResult, Ref, Message } from '../types';
 
 export default function getValidateError(
   result: ValidateResult,
   ref: Ref,
   type = 'validate',
 ): FieldError | void {
-  const isStringValue = isString(result);
+  const isMessage = (value: any): value is Message =>
```
Comment (bluebill1049) [inline src/logic/getValidateError.ts]: super minor: any chance we can avoid `any`. (try to challenge you hahaha I know I would have used `any`) 😂

Comment (bluebill1049) [reply src/logic/getValidateError.ts]: also what about turn this function in a separate file? 

```
import isMessageFunction from './utls/isMessage';

const isMessage = isMessageFunction(value)
```

we can avoid running the function twice below and a little bit cleaner.

File: src/logic/getValueAndMessage.ts [hunk 3]
```diff
@@ -1,26 +1,15 @@
...
+    isObject(value) && !isRegex(value);
 
   return {
-    value: isPureObject
-      ? (validationData as ValidationValueMessage).value
-      : (validationData as ValidationValue),
-    message: isPureObject
-      ? (validationData as ValidationValueMessage).message
-      : '',
+    value: isValueMessage(validationData)
```
Comment (bluebill1049) [inline src/logic/getValueAndMessage.ts]: wow nice! you remove all my rubbish `as`

File: src/logic/validateField.test.tsx [hunk 4]
```diff
@@ -862,5 +1235,25 @@ describe('validateField', () => {
...
+          value: /d/i,
+          message: <p>pattern</p>,
+        },
+        validate: {
+          test: value => value === <p>test</p>,
+          test1: value => value == 'test' || <p>Luo</p>,
+          test2: value => value == 'test' || <p>Bill</p>,
+        },
+      }),
+    ).toMatchSnapshot();
```
Comment (bluebill1049) [inline src/logic/validateField.test.tsx]: Thank you for those extra tests! <3

File: src/logic/validateField.ts [hunk 5]
```diff
@@ -74,11 +74,10 @@ export default async <FormValues extends FieldValues>(
       (isCheckBox && !getCheckboxValue(options).isValid) ||
       (isRadio && !getRadioValue(options).isValid))
   ) {
-    const { value: requiredValue, message: requiredMessage } = isString(
-      required,
-    )
-      ? { value: !!required, message: required }
-      : getValueAndMessage(required);
+    const { value: requiredValue, message: requiredMessage } =
+      isString(required) || React.isValidElement(required)
```
Comment (bluebill1049) [inline src/logic/validateField.ts]: I think we can reuse that `isMessage` function here.

File: src/types.ts [hunk 6]
```diff
@@ -92,21 +92,32 @@ export type MutationWatcher = {
...
 };
 
-type ValidationOptionObject<Value> = Value | { value: Value; message: string };
+export type Message = string | React.ReactElement;
+
+export type ValidationOption<Value extends ValidationValue = ValidationValue> =
+  | Value
+  | ValidationValueMessage<Value>;
 
 export type ValidationValue = boolean | number | string | RegExp;
```
Comment (bluebill1049) [inline src/types.ts]: Maybe move `ValidationValue` on top of `ValidationOption`?

File: src/types.ts [hunk 7]
```diff
@@ -92,21 +92,32 @@ export type MutationWatcher = {
...
 
-export type ValidateResult = string | boolean | undefined;
+export type ValidationValueMessage<
```
Comment (bluebill1049) [inline src/types.ts]: Love this. 🏅

**Review by bluebill1049 — COMMENTED** [review]
Quality Quality Quality!

Amazing work and beyond! Thank you @kotarella1110 (just some minor comments)

File: src/logic/getValidateError.ts (hunk 2 above)
Comment (kotarella1110) [reply src/logic/getValidateError.ts]: @bluebill1049 I agree with it 👍 I will fix!

File: src/logic/validateField.ts (hunk 5 above)
Comment (kotarella1110) [reply src/logic/validateField.ts]: Oh nice! I will 👍

File: src/types.ts (hunk 6 above)
Comment (kotarella1110) [reply src/types.ts]: I will it!

File: src/logic/getValidateError.ts (hunk 2 above)
Comment (kotarella1110) [reply src/logic/getValidateError.ts]: > we can avoid running the function twice below and a little bit cleaner.

@bluebill1049 From a typeguard point of view I want to execute this function twice.

スクリーンショット 2020-03-26 21 44 15

スクリーンショット 2020-03-26 21 45 37

Comment (kotarella1110) [reply src/logic/getValidateError.ts]: > https://github.com/react-hook-form/react-hook-form/pull/1275#discussion_r398487266

completed!

File: src/logic/validateField.ts (hunk 5 above)
Comment (kotarella1110) [reply src/logic/validateField.ts]: completed!

File: src/types.ts (hunk 6 above)
Comment (kotarella1110) [reply src/types.ts]: completed!

**Review by bluebill1049 — APPROVED** [review]
Awesome! can't wait for this to go live.

**Review by bluebill1049 — APPROVED** [review]
🎆🎇🎆🎇🎆🎇

**Comment by kotarella1110** [comment]
@bluebill1049 Thank you for your review 💚

**Comment by kotarella1110** [comment]
@bluebill1049 I will update the document later 👍

**Comment by bluebill1049** [comment]
awesome @kotarella1110!

**Pull request merged**